src/_pytest/pathlib.py

```diff
@@ -689,6 +733,48 @@ def resolve_package_path(path: Path) -> Optional[Path]:
     return result
 
 
+def resolve_pkg_root_and_module_name(
+    path: Path, *, consider_ns_packages: bool = False
+) -> Tuple[Path, str]:
+    """
+    Return the path to the directory of the root package that contains the
+    given Python file, and its module name:
+
+        src/
+            app/
+                __init__.py
+                core/
+                    __init__.py
+                    models.py
+
+    Passing the full path to `models.py` will yield Path("src") and "app.core.models".
+
+    If consider_ns_packages is True, then we additionally check if the top-level directory
```

Comment:
1) Yeah that's something we can revisit later, pytest never supported packages without a `__init__` correctly (is there a name for it?), I recall we had several discussions with other package-related authors (pip maintainers, Jason Coombs, etc) on how to support that, but the discussion stalled. I would be happy to tackle that in a separate PR though, as this requires more tests in order to make sure the support is sound.

2) Do they? I always thought it was a mechanism just to share the top-level, based on the use cases out there and the docs at https://packaging.python.org/en/latest/guides/packaging-namespace-packages/#native-namespace-packages. I will double check if this is really supported, and if so update the code accordingly.

Reply:
About 2, you are right, it does support multiple levels (TIL). :+1:

Reply:
I'm hesitant on adding multiple level support for namespace packages, let me explain my reasoning.

I can see situations (might be contrived I admit) where we identify a module as part of a namespace package incorrectly, just because the directory is in `sys.path` somehow.

For example:


```
temp/
  some/
    folder/
      myproject/
        src/
          com/
            company/
              myapp/
                __init__.py
                core.py
```

If we look upwards from `core.py` in order to find if it is in a namespace package, we must keep looking upwards until we find a directory in `sys.path` or until we reach the root of the file system and give up. Suppose for some reason `temp/` is in `sys.path` (possibly due to a misconfiguration), this means we could end up determining that `core.py` is be part of the package `temp.some.folder.myproject.src.com.company.myapp.core`, which is incorrect, depending on the order directories appear in  `sys.path`, if `temp/` appears first: `temp/` or `temp/some/folder/myproject/src`.

The namespace package rules are easy to follow when you try to figure out a *file path* starting from a *module name*: to see if `com.company.myapp.core` as it appears in the source code is valid , we need to through each `sys.path` entry, join it with `com/company/myapp/core.py` and see if the file exists: if it does, double check that there are no `__init__.py` files present, then import it. This is what Python needs to do and why it works unambiguously.

But the situation is not so simple when you need to derive a *module name* starting from a *file path*, because you need to search recursively upwards and see if you can build a valid module name using that full path, but even if we do build a valid module name we don't know if that was intentional or not, and we might end up with a module named `temp.some.folder.myproject.src.com.company.myapp.core`; in the previous case, going from module name to path, the user is telling us the module name, so there is no ambiguity.

I recall that was the central point  when this matter was being discussed with other package-related authors in the past (I cannot find the thread, perhaps @RonnyPfannschmidt can recall at least which forum that took place?).

Hope I was able to explain the situation, sorry if the explanation is lacking.

Btw, I only introduced this change regarding namespace packages because 1 specific test in doctests that is meant to import from namespace packages, but AFAICT it was working with namespace packages by accident, due to how we were building the module name from the path, not due to any proper namespace package support -- so much that I changed it slighly the test to use a `src` layout, while keeping the namespace package configuration, and then the code in `main` could not find it anymore. The change implemented here made we at least recognize namespace packages with one level.

I might be overthinking this too, so would like to hear your opinions on this.

Unless my assessment is incorrect, I think a compromise would be to:

1. Support just one level for now; up until now we did not support any level at all, so this is a small test to provide full support.
2. Keep the status quo of "pytest does not yet support namespace packages", remove the partial support I added here, and mark the test as xfail.
3. Implement full support in this PR.

I'm OK with 3), just wanted to present the full picture and see if this is the direction we want to go.

There might also a simpler/safer way to determine if a file path is part of a full namespace package that I'm not seeing, in which case much of my concerns here are moot.

Reply:
To aid in the discussion I went ahead and implemented it in 1232d198ae5e.

The implementation is simple, I'm a bit hesitant if this can have undesired side effects in the wild.

Reply:
I think your analysis is correct, particularly the difference between going module path => filesystem path (what Python does) and the reverse (what we want to do).

I think the partial support is too confusing, so IMO we shouldn't do option 1 (I prefer 2 over 1).

I totally understand the concern with option 3. However, I think we can try it and see. We may end up with these long unintended module paths, but I don't think it would be actively *dangerous* (=> actually importing unintended stuff) because intermediate paths in namespace packages are not executing anything (there is no `__init__.py` to import).

Longer term, and this has already been brought by several people, I do think we need to add some concept of "workspace roots", which are `Directory` collector nodes directly below `Session`, and which put a strict bound on filesystem traversals (never going beyond them). (Also, initial conftests only collected from workspace roots. The `confcutdir` concept is gone.  `--pypkg` args become workspace roots. All nodeids are rooted in roots, can't have `<rootdir>/../bla.py` or `/foo/bar.py` nodeids anymore). Hopefully I can put up a proper proposal sometime soon.

---

> Suppose for some reason temp/ is in sys.path (possibly due to a misconfiguration), this means we could end up determining that core.py is be part of the package temp.some.folder.myproject.src.com.company.myapp.core, which is incorrect, depending on the order directories appear in sys.path, if temp/ appears first: temp/ or temp/some/folder/myproject/src.

Since we're iterating `pkg_path.parents` in order, I don't think there is ambiguity -- we'll always try  `com.company.myapp.core` in `temp/some/folder/myproject/src` first, so that's the one that would be found, even if `temp/` is in sys.path.

---

Regarding the initial implementation, in the loop `for parent in pkg_path.parents:`, is there some special handling needed if one of the `parents` has a `__init__.py`? Can namespace packages be nested in regular packages? Or the other way around?

Reply:
> Since we're iterating pkg_path.parents in order, I don't think there is ambiguity -- we'll always try com.company.myapp.core in temp/some/folder/myproject/src first, so that's the one that would be found, even if temp/ is in sys.path.

Yes, you are right.


> Regarding the initial implementation, in the loop for parent in pkg_path.parents:, is there some special handling needed if one of the parents has a __init__.py? 

Good catch, indeed if there is an `__init__.py` it is no longer a namespace package. Added a specific test for this.

---

Went ahead with option 3 then, please take a look.

I also double checked and the `pytest-doctest-import-mismatch` (the reproducer for the issue) still passes.

Reply:
> Good catch, indeed if there is an __init__.py it is no longer a namespace package. Added a specific test for this.

So I've just seen in the link you provided https://packaging.python.org/en/latest/guides/packaging-namespace-packages/#native-namespace-packages a regular package nested inside of a namespace package... So I assume it is possible?